### PR TITLE
Feat: implement automatic re-generation of models

### PR DESCRIPTION
**Pull request by nguyentoanit**

### CODE REVIEW

**Comment by moltar** [comment]
Task linked: [GHORG-17 Implement automatic re-generation of models](https://app.clickup.com/t/2250003/GHORG-17)

File: gulpfile.ts [hunk 1]
```diff
@@ -1,5 +1,9 @@
+import SwaggerParser from '@apidevtools/swagger-parser' // eslint-disable-line import/no-extraneous-dependencies
```
Comment (moltar) [inline gulpfile.ts:1]: To fix the `import/no-extraneous-dependencies` rule, please make a PR to the `utils` repo, and add `gulpfile.ts` to the array of file names:

https://github.com/ScaleLeap/utils/blob/5782c814ba042c019ab65957f26a89051cc9021e/eslint/index.js#L164

Once `utils` package release, you can update it and the warning will be gone.

File: gulpfile.ts [hunk 2]
```diff
@@ -14,12 +18,17 @@ type GithubObject = GetType<typeof GithubObject>
...
+  outputPath: string
 }
 
 // Using authentication to increases Github API rate limit.
 const octokit = new Octokit()
 const owner = 'amzn'
 const repo = 'selling-partner-api-models'
+const redundantFiles: string[] = ['.gitignore', '.openapi-generator-ignore', 'git_push.sh']
+const redundantDirectories: string[] = ['.openapi-generator']
+const exclusionExportingObjects = new Set(['ErrorList', 'Error'])
```
Comment (moltar) [inline gulpfile.ts:31]: Please rename to `excludeExportedObjects`

File: gulpfile.ts [hunk 3]
```diff
@@ -36,52 +45,108 @@ async function fetchContentsByPath(repoPath = 'models'): Promise<GithubObject[]>
...
    *- git_push.sh
    */
 
-  return model
+  return Promise.all([
+    ...redundantFiles.map((object) => fsPromises.unlink(`${model.outputPath}/${object}`)),
+    ...redundantDirectories.map((object) =>
+      fsPromises.rmdir(`${model.outputPath}/${object}`, { recursive: true }),
+    ),
+  ]).then((_) => model) // eslint-disable-line @typescript-eslint/no-unused-vars
```
Comment (moltar) [inline gulpfile.ts:83]: Why do we need to use `then`?

Why not:

```ts
await Promise.all([...])

return model
```

File: gulpfile.ts [hunk 4]
```diff
@@ -36,52 +45,108 @@ async function fetchContentsByPath(repoPath = 'models'): Promise<GithubObject[]>
...
+}
+
+function verifyObjectDefinition(definitions: Record<string, any>, key: string): boolean {
```
Comment (moltar) [inline gulpfile.ts:86]: It is not clear what this function does. Please add some comments why we need it and what it does.

File: gulpfile.ts [hunk 5]
```diff
@@ -36,52 +45,108 @@ async function fetchContentsByPath(repoPath = 'models'): Promise<GithubObject[]>
...
 
-  return githubFiles
+  const apiModelGeneratorPromises = githubFiles
     .flat()
     .map(generateAPIModel)
     .map(executeGeneratorCLI)
-    .map(removeRedundantObjects)
-    .map(exportAPIModel)
+
+  await Promise.all<APIModel>(apiModelGeneratorPromises)
```
Comment (moltar) [inline gulpfile.ts:144]: Not clear why we need to mix `await` and `then`.

Can we use `await` for every call here?

File: gulpfile.ts [hunk 6]
```diff
@@ -14,12 +18,17 @@ type GithubObject = GetType<typeof GithubObject>
...
   command: string
+  modelName: string
+  outputPath: string
 }
 
 // Using authentication to increases Github API rate limit.
 const octokit = new Octokit()
 const owner = 'amzn'
 const repo = 'selling-partner-api-models'
+const redundantFiles: string[] = ['.gitignore', '.openapi-generator-ignore', 'git_push.sh']
```
Comment (moltar) [inline gulpfile.ts:29]: All truly constant names should use constant case:

```
repo => REPO
redundantFiles => REDUNDANT_FILES
```

**Review by moltar — CHANGES_REQUESTED** [review]

File: gulpfile.ts (hunk 5 above)
Comment (nguyentoanit) [reply gulpfile.ts:144]: @moltar : Something like this?

```
  const apiModels = await Promise.all<APIModel>(apiModelGeneratorPromises)
  await Promise.all([
    ...apiModels.map(removeRedundantObjects),
    apiModels.map(generateExportStatement).then(writeStatementsToFile),
  ])
```

Comment (moltar) [reply gulpfile.ts]: Refactor `.then(writeStatementsToFile)` too :)

File: gulpfile.ts [hunk 7]
```diff
@@ -36,52 +45,124 @@ async function fetchContentsByPath(repoPath = 'models'): Promise<GithubObject[]>
...
+        exportings.push(`${key} as ${upperFirst(camelCase(model.modelName))}${key}`)
+      }
+    }
+
+    return `export { ${exportings.join(', ')} } from './${model.modelName}'`
+  })
+}
+
+function writeStatementsToFile(statements: string[]): void {
+  return fs.writeFileSync('src/api-models/index.ts', statements.join('\n'))
```
Comment (moltar) [inline gulpfile.ts:146]: Why is this sync when everything else is async?

Comment (nguyentoanit) [reply gulpfile.ts:146]: @moltar : I don't want to await an async function while I can make it to sync. 
And performance isn't also optimized when I use async in there. 😕

**Review by moltar — APPROVED** [review]

**Pull request merged**